### PR TITLE
d.ts improvements

### PR DESCRIPTION
**Pull request by valotas**
- Introduced a `FunctionalComponent` type to describe functional components
- Added context to the Component's interfaces

### CODE REVIEW

File: src/preact.d.ts [hunk 1]
```diff
@@ -33,18 +33,23 @@ declare namespace preact {
 		componentDidUpdate?(previousProps:PropsType,previousState:StateType,previousContext:any):void;
 	}
 
+	interface FunctionalComponent<PropsType, StateType> {
```
Comment (developit) [inline src/preact.d.ts]: I don't know enough about TS to say for sure, but I think this should context? The signature for Pure Functional Components is `fn(props, context) => jsx`

Comment (valotas) [inline src/preact.d.ts]: You are right. I'll fix that

File: src/preact.d.ts [hunk 2]
```diff
@@ -57,7 +62,7 @@ declare namespace preact {
 		abstract render(props:PropsType & ComponentProps<this>, state:any):JSX.Element;
```
Comment (valotas) [inline src/preact.d.ts]: I believe here I should also add the context as a third argument, right?

Comment (developit) [inline src/preact.d.ts]: yup!

File: src/preact.d.ts (hunk 1 above)
Comment (developit) [inline src/preact.d.ts]: sounds good to me :)

**Comment by valotas** [comment]
I'll try to add some typescript compilation tests too

**Comment by developit** [comment]
@valotas that would be *awesome*!

**Comment by valotas** [comment]
@developit I think that this can be merged. Do you need anything else from me?

**Comment by developit** [comment]
Nope, this all looks great to me. Would love a quick once-over by someone with more TS knowledge, I am mostly clueless about TS.

**Comment by NekR** [comment]
I haven't used TS for a while tbh. Seems fine, but TS declaration format evolved from last time I used TS.

File: test/ts/prect-test.tsx [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import { h, render, Component } from 'preact';
```
Comment (robertknight) [inline test/ts/prect-test.tsx]: Typo in the name of this file "prect-test"

File: test/ts/tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,24 @@
+{
+	"compilerOptions": {
+		"module": "commonjs",
+		"lib": [
+			"es6",
+			"dom"
+		],
+		"noImplicitAny": true,
+		"noImplicitThis": true,
+		"strictNullChecks": false,
```
Comment (robertknight) [inline test/ts/tsconfig.json]: Out of interest, would the tests pass if `strictNullChecks` were enabled?

**Pull request merged**